sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/JoinEstimationSuite.scala

```diff
@@ -0,0 +1,356 @@
+      // For now we just propagate the statistics from left side for left semi/anti join.
+      val expectedStats = Statistics(
+        sizeInBytes = 3 * (8 + 4 * 2),
+        rowCount = Some(3),
+        attributeStats = AttributeMap(Seq(nameToColInfo("key-1-2"), nameToColInfo("key-2-4"))))
+      assert(join.stats(conf) == expectedStats)
+    }
+  }
+
+  test("test join keys of different types") {
+    val dec1 = new java.math.BigDecimal("1.000000000000000000")
+    val dec2 = new java.math.BigDecimal("8.000000000000000000")
+    val d1 = Date.valueOf("2016-05-08")
+    val d2 = Date.valueOf("2016-05-09")
+    val t1 = Timestamp.valueOf("2016-05-08 00:00:01")
+    val t2 = Timestamp.valueOf("2016-05-09 00:00:02")
+
+    /** Columns in a table with only one row */
+    val columnInfo1 = mutable.LinkedHashMap[Attribute, ColumnStat](
+      AttributeReference("cbool", BooleanType)() -> ColumnStat(distinctCount = 1,
+        min = Some(false), max = Some(false), nullCount = 0, avgLen = 1, maxLen = 1),
+      AttributeReference("cbyte", ByteType)() -> ColumnStat(distinctCount = 1,
+        min = Some(1L), max = Some(1L), nullCount = 0, avgLen = 1, maxLen = 1),
+      AttributeReference("cshort", ShortType)() -> ColumnStat(distinctCount = 1,
+        min = Some(1L), max = Some(1L), nullCount = 0, avgLen = 2, maxLen = 2),
+      AttributeReference("cint", IntegerType)() -> ColumnStat(distinctCount = 1,
+        min = Some(1L), max = Some(1L), nullCount = 0, avgLen = 4, maxLen = 4),
+      AttributeReference("clong", LongType)() -> ColumnStat(distinctCount = 1,
+        min = Some(1L), max = Some(1L), nullCount = 0, avgLen = 8, maxLen = 8),
+      AttributeReference("cdouble", DoubleType)() -> ColumnStat(distinctCount = 1,
+        min = Some(1.0), max = Some(1.0), nullCount = 0, avgLen = 8, maxLen = 8),
+      AttributeReference("cfloat", FloatType)() -> ColumnStat(distinctCount = 1,
+        min = Some(1.0), max = Some(1.0), nullCount = 0, avgLen = 4, maxLen = 4),
+      AttributeReference("cdecimal", DecimalType.SYSTEM_DEFAULT)() -> ColumnStat(distinctCount = 1,
+        min = Some(dec1), max = Some(dec1), nullCount = 0, avgLen = 16, maxLen = 16),
+      AttributeReference("cstring", StringType)() -> ColumnStat(distinctCount = 1,
+        min = None, max = None, nullCount = 0, avgLen = 3, maxLen = 3),
+      AttributeReference("cbinary", BinaryType)() -> ColumnStat(distinctCount = 1,
+        min = None, max = None, nullCount = 0, avgLen = 3, maxLen = 3),
+      AttributeReference("cdate", DateType)() -> ColumnStat(distinctCount = 1,
+        min = Some(d1), max = Some(d1), nullCount = 0, avgLen = 4, maxLen = 4),
+      AttributeReference("ctimestamp", TimestampType)() -> ColumnStat(distinctCount = 1,
+        min = Some(t1), max = Some(t1), nullCount = 0, avgLen = 8, maxLen = 8)
+    )
+
+    /** Columns in a table with two rows */
+    val columnInfo2 = mutable.LinkedHashMap[Attribute, ColumnStat](
```

Comment:
are these totally same with the `columnInfo1`? may we can create a method to do this